### PR TITLE
Clef changes

### PR DESCRIPTION
**Pull request by massimobio**
This is a PR is branched off of the clefPerNote branch and related #118 
The goal is to complete all discussed clef changes. I'm using a separate branch as discussed to make sure this is the direction we want to go in.
Removed all clef properties and related methods in Note and NoteCollection.
@drumnkyle has already implemented clefForNoteAtIndex, clefForNoteCollectionAtIndex and clefForTicks in Measure so we have all the necessary APIs to determine the clef for any given note, and this should also work for tuplets since we can ask for the clef at any arbitrary tick position.
Tests remain to be added for the new APIs.
If we want to continue in this direction I can write more tests.
Anything else?

### CODE REVIEW

**Comment by drumnkyle** [comment]
Just reviewed the changes you made so far. Seems good. You can continue. I read over our notes in the OneNote as well. Make sure you review those as there are multiple things to do. Also, there is a checklist on the original PR.

**Comment by massimobio** [comment]
As per one of the remaining todo's in the original PR, I added a simple debug description to measure, that includes the clef info, but it doesn't look that great:
```Test Case '-[MusicNotationCoreMacTests.MeasureTests testChangeClefInMiddleHasBeginningAndEnd]' started.
|4/4: [1/16c1, 1/16c1, 1/16c1, 1/16c1],[1/8c1, 1/8c1, 1/8c1, 1/8c1, 1/8c1, 1/8c1, 1/8c1, 1/8c1]| [0: treble, 3072: bass, 7168: treble]
```
A prettier version would show clef changes before the note at which they happen. Tried playing around with this a bit but it's a bigger job involving check each note's cumulative tick and comparing with clefs and I got bogged down with sets and NoteCollections... 
Since it doesn't affect the API directly we could perhaps make this a separate Issue to be dealt with later?

**Comment by massimobio** [comment]
Cool, I made a separate issue for that.

Question regarding a test stub you had written:
```func testChangeClefAtBeginningAlreadyThere()```
What do you mean by "AlreadyThere"?

for reference here is another related test:
```
    func testChangeClefInMiddleHasBeginning() {
        let eighth = Note(noteDuration: .eighth, tone: Tone(noteLetter: .c, octave: .octave1))
        let quarter = Note(noteDuration: .sixteenth, tone: Tone(noteLetter: .c, octave: .octave1))
        var measure = Measure(timeSignature: timeSignature, notes: [
            [
                quarter, quarter, quarter, quarter
            ],
            [
                eighth, eighth, eighth, eighth, eighth, eighth, eighth, eighth
            ]
            ])
        assertNoErrorThrown {
            try measure.changeClef(.treble, at: 0, inSet: 1)
            try measure.changeClef(.bass, at: 3, inSet: 1)
            XCTAssertEqual(measure.clefs, [0: .treble, 3072: .bass])
            XCTAssertEqual(measure.lastClef, .bass)
            XCTAssertEqual(measure.originalClef, nil)
        }
    }
```

**Comment by drumnkyle** [comment]
Hmm. My guess is that it means make a Clef change at the beginning (0) and then change the Clef again at 0.

**Comment by massimobio** [comment]
Interesting. The second entry should override the first. I'll play with that idea.

**Comment by drumnkyle** [comment]
I believe so. Btw just do reference later, the new issue you made is #126. Just figure it’s good to link these two together.

**Comment by massimobio** [comment]
I have fixed all clef related tests and also fixed some bugs in calculating cumulative ticks for tuplets.
There remains 3 failing tests not related to Clefs. One is due to a decimal rounding error for tuples with a ratio that doesn't fit evenly in the 2048 ticks/quarter system. Not sure what to do about that but we have failing tests in place and it's an edge case and not related to the clef so we should probably discuss a fix for this separately. It might also require some major changes.

The other failing test is for compound tuplets testCumulativeTicksAtBeginningOfCompoundTuplet.
I am a little confused about the compound tuplets:
```
let note = Note(noteDuration: .eighth)
measure.append(note)
let triplet = try Tuplet(3, .eighth, notes: [note, note, note])
let compoundTuplet = try Tuplet(5, .eighth, notes: [note, note, triplet, note])
measure.append(compoundTuplet)
print(measure.debugDescription)
```
The debug print for the measure is  ``` // |4/4: [1/8R, 6[1/8R, 1/8R, 3[1/8R, 1/8R, 1/8R], 1/8R]]|```
Why does it say 6? should it be 5?
I'm trying to figure out the correct ticks so I can write the correct test

In any case all these edge case issues have failing tests in place and are not directly related to the clef changes in this PR. Perhaps we can review everything else in this PR and if it looks right we can merge since this is the hold up for several other PR's and issues.

**Comment by drumnkyle** [comment]
I agree with everything you said. Are you saying your PR is ready to be fully reviewed and merged? We can work on that issue in new PR. 

As far as the compound tuplets, yeah that’s weird. I’m quite sure it should print out 5. Seems like a bug in the debug description. The setup you did seems right to me.

**Comment by massimobio** [comment]
I got Travis to work again. It was still using Xcode 8.
Also added another failing test for some code that wasn't covered.
Won't touch anything else now.

**Comment by drumnkyle** [comment]
Ok. I'm going to start reviewing everything now.

**Comment by massimobio** [comment]
Yeah, the failing tests are all related to tuplets. I think you added new methods to the tuplets to calculate the ticks but some of them need some work.
screen shot 2018-08-14 at 12 26 35 am

File: Sources/Measure.swift [hunk 1]
```diff
@@ -309,6 +410,7 @@ public struct Measure: ImmutableMeasure, Equatable, RandomAccessCollection {
         - `MeasureError.invalidTieState`
      */
     internal mutating func insert(_ noteCollection: NoteCollection, at index: Int, inSet setIndex: Int = 0, shouldIgnoreTieStates skipTieConfig: Bool) throws {
+
```
[the remark this answers is not in the document]
Comment (massimobio) [reply Sources/Measure.swift]: Agreed, wil fix.

File: Sources/Measure.swift [hunk 2]
```diff
@@ -755,13 +900,49 @@ public struct Measure: ImmutableMeasure, Equatable, RandomAccessCollection {
     private func tieState(for index: Int, inSet setIndex: Int) throws -> Tie? {
         return try note(at: index, inSet: setIndex).tie
     }
+
+    /**
+     Returns the number of ticks that exist in the measure up to, but not including
+     the given note index.
+     */
+    internal func cumulativeTicks(at noteIndex: Int, inSet setIndex: Int = 0) throws -> Int {
+        // Total # of ticks up to, but not including the given noteIndex
```
[the remark this answers is not in the document]
Comment (massimobio) [reply Sources/Measure.swift]: Will remove.

File: Tests/MusicNotationCoreTests/StaffTests.swift [hunk 3]
```diff
@@ -144,33 +150,37 @@ class StaffTests: XCTestCase {
 			let addedMeasure = try staff.measure(at: 14)
 			let beforeMeasure = try staff.measure(at: 13)
 			let afterMeasure = try staff.measure(at: 15)
-			XCTAssertEqual(Measure(addedMeasure), measure)
-			XCTAssertEqual(Measure(beforeMeasure), measure6)
-			XCTAssertEqual(Measure(afterMeasure), measure3)
+            XCTAssertEqual(Measure(addedMeasure), changedClef(of: measure))
```
[the remark this answers is not in the document]
Comment (massimobio) [reply Tests/MusicNotationCoreTests/StaffTests.swift]: Yeah, This methods used tabs instead of spaces for some reason. Will reformat the whole method.

**Comment by drumnkyle** [comment]
Aren't those new methods to calculate the ticks needed for the clef changes to work though? Doesn't that mean that this change isn't really finished until those are fixed?

**Comment by massimobio** [comment]
That's correct but the fixes aren't trivial. 
A group of failing tests are due to wrong tick count in certain tuplets because of decimal point rounding. The other issue has to do with some bug in creating compound tuplets as I described above which is not new to this branch.

**Comment by drumnkyle** [comment]
Gotcha. OK. I guess it makes sense to merge this to get this in so we can do some upgrades to Swift 4 and stuff. Then, we can tackle the other changes. I'm not a big fan of breaking master, but since we shouldn't really have any consumers at this point and we aren't going to bump the Cocoapods version right now, I think we should be fine.

**Comment by drumnkyle** [comment]
Shall I merge?

**Comment by massimobio** [comment]
The bugs are for changing the clef in the middle of tuplets and compound tuplets so I doubt this will affect anybody right now.
I can change the failing tests to FIXMEs if you prefer. That way Travis would succeed and we can better detect new bugs

**Comment by drumnkyle** [comment]
I mean, Travis is already failing right and is not being detected. So, if it is actually failing, we should probably leave it that way I guess, right? It should be a top priority to fix the tests so that we can have everything passing. I don't want to leave it in a failing state for long. However, I'm ok with leaving it in a failing state while we do some cleanup and upgrade to Swift 4. 

I'll merge now.
